File: programmers/level3/가장_먼_노드.java
import java.util.*;
class Solution {
	HashMap<Integer, HashSet<Integer>> map = new HashMap<Integer, HashSet<Integer>>();
	int answer = 0;
	public int solution(int n, int[][] edge) {
        boolean[] visited = new boolean[n+1];
        for(int i = 0; i < edge.length; ++i) {
        	if(!map.containsKey(edge[i][0])) {
        		map.put(edge[i][0],new HashSet<Integer>());
        	}
        	map.get(edge[i][0]).add(edge[i][1]);
        	if(!map.containsKey(edge[i][1])) {
        		map.put(edge[i][1],new HashSet<Integer>());
        	}
        	map.get(edge[i][1]).add(edge[i][0]);
        }
        bfs(n,edge,visited);
        return answer;
    }
    void bfs(int n, int[][] edge, boolean[] visited) {
    	int current = 1;
    	visited[1] = true;
    	Queue<Integer> q = new LinkedList<Integer>();
    	q.add(current);
    	while(!q.isEmpty()) {
    		int size = q.size();
   			answer = size;
    		for(int i = 0; i < size; ++i) {
    			current = q.remove();
    			for(int next : map.get(current)) {
    				if(!visited[next]) {
    					visited[next] = true;
    					q.add(next);
    				}
    			}
    		}
    	}
    }
}
public class 가장_먼_노드 {
	public static void main(String[] args) {
		int n = 6;
		int[][] vertex = {{3, 6}, {4, 3}, {3, 2}, {1, 3}, {1, 2}, {2, 4}, {5, 2}};
		Solution sol = new Solution();
		System.out.println(sol.solution(n, vertex));
	}
}

File: programmers/js/level3/가장_먼_노드.js
function solution(n, edge) {
    const map = new Map();
    edge.forEach(row => {
        let f = row[0];
        let t = row[1];
        if (map.get(f) == undefined) {
            map.set(f, new Set());
        }
        if (map.get(t) == undefined) {
            map.set(t, new Set());
        }
        map.get(f).add(t);
        map.get(t).add(f);
    });

    const q = [];
    let current = 1;
    q.push(current);
    let answer = 0;
    let visited = Array(n + 1);
    for (let i = 1; i <= n; ++i) {
        visited[i] = false;
    }
    visited[1] = true;
    while (q.length > 0) {
        let size = q.length;
        answer = size;
        for (let i = 0; i < size; ++i) {
            current = q.shift();
            const s = map.get(current);
            s.forEach(t => {
                if (!visited[t]) {
                    visited[t] = true;
                    q.push(t);
                }
            })
        }
    }
    return answer;
}

console.log(solution(6, [[3, 6], [4, 3], [3, 2], [1, 3], [1, 2], [2, 4], [5, 2]]));